common/pose.h

```diff
@@ -1,23 +1,209 @@
 #pragma once
 
-// Standard C++ includes
-#include <array>
-
 // Third-party includes
 #include "../third_party/units.h"
 
+// Standard C++ includes
+#include <array>
+#include <tuple>
+
 namespace BoBRobotics {
-//! A generic template for 2D unit arrays
-template<class T>
-using Vector2 = std::array<T, 2>;
 
-//! A generic template for 3D unit arrays
-template<class T>
-using Vector3 = std::array<T, 3>;
+//! A generic template for 2D arrays
+template<typename T>
+using Array2 = std::array<T, 2>;
+
+//! A generic template for 3D arrays
+template<typename T>
+using Array3 = std::array<T, 3>;
+
+//! Base class for vectors of length units
+template<typename LengthUnit, size_t N>
+class VectorBase
+{
+    static_assert(units::traits::is_length_unit<LengthUnit>::value,
+                  "LengthUnit is not a unit of length");
+
+public:
+    VectorBase() = default;
+
+    template<typename... Ts>
+    VectorBase(Ts &&... args)
+      : m_Array({ std::forward<Ts>(args)... })
+    {}
+
+    operator const std::array<LengthUnit, N> &() const
+    {
+        return m_Array;
+    }
+
+    LengthUnit &operator[](size_t i) { return m_Array[i]; }
+    const LengthUnit &operator[](size_t i) const { return m_Array[i]; }
+    static constexpr size_t size() { return N; }
+
+    auto begin() { return m_Array.begin(); }
+    auto begin() const { return m_Array.begin(); }
+    auto end() { return m_Array.end(); }
+    auto end() const { return m_Array.end(); }
+    auto cbegin() const { return m_Array.cbegin(); }
+    auto cend() const { return m_Array.end(); }
+
+private:
+    std::array<LengthUnit, N> m_Array;
+};
+
+template<typename LengthUnit>
+class Vector3;
+
+//! 2D length unit vector
+template<typename LengthUnit>
+class Vector2
+  : public VectorBase<LengthUnit, 2>
+{
+public:
+    Vector2() = default;
+
+    Vector2(LengthUnit x, LengthUnit y)
+      : VectorBase<LengthUnit, 2>(x, y)
+    {}
+
+    Vector2(const std::array<LengthUnit, 2> &array)
+      : Vector2(array[0], array[1])
+    {}
+
+    operator Vector3<LengthUnit>() const { return Vector3<LengthUnit>(x(), y(), z()); }
+
+    LengthUnit &x() { return (*this)[0]; }
+    const LengthUnit &x() const { return (*this)[0]; }
+    LengthUnit &y() { return (*this)[1]; }
+    const LengthUnit &y() const { return (*this)[1]; }
+    static constexpr LengthUnit z() { return LengthUnit(0); }
+};
+
+//! 3D length unit vector
+template<typename LengthUnit>
+class Vector3
+  : public VectorBase<LengthUnit, 3>
+{
+public:
+    Vector3() = default;
+
+    Vector3(LengthUnit x, LengthUnit y, LengthUnit z)
+      : VectorBase<LengthUnit, 3>(x, y, z)
+    {}
+
+    Vector3(const std::array<LengthUnit, 3> &array)
+      : Vector3(array[0], array[1], array[2])
+    {}
+
+    operator Vector2<LengthUnit>() const { return Vector2<LengthUnit>(x(), y()); }
```

Comment:
Sure, that sounds good. I'm guessing you mean we could return ``std::array``s from this rather than a bunch of new classes though....?

Reply:
Does it make any difference? those classes are only a very thin wrapper over ``std::array`` anyway - it shouldn't make any difference performance-wise

Reply:
I wasn't thinking about performance: it's just that a ``Vector2`` treats the two members of its array as x and y. So it's probably easier to return ``std::array``s rather than making some new classes for it.